titus_isolate/event/oversubscribe_event_handler.py

```diff
@@ -0,0 +1,280 @@
+        pred_env = PredEnvironment(self.__config_manager.get_region(), self.__config_manager.get_environment(),
+                                   datetime.utcnow().hour)
+
+        if self.__is_window_active():
+            self.__skip_count += 1
+            self.handled_event(event, 'skipping oversubscribe - a window is currently active')
+            return
+
+
+        workload_count = 0
+        free_cpu_count = 0
+        # we calculate the window before we send the request to ensure we're not going over our 10 minute mark
+        start = datetime.utcnow()
+        end = start + timedelta(minutes=10)
+        for workload in self.workload_manager.get_workloads():
+            # skip if an opportunistic workload already exists on this agent. this case only happens when an
+            # opportunistic workload outlives the window it was created for.
+            # TODO optimize by subtracting opportunistic CPUs used by opportunistic workloads instead of skipping.
+            # TODO optimize by reclaiming unused real CPUs from partially opportunistic workloads.
+            if workload.is_opportunistic():
+                self.__skip_count += 1
+                self.handled_event(event, 'skipping oversubscribe - an opportunistic workload already exists')
+                return
+
+            log.debug('workload:%s job_type:%s cpu:%d', workload.get_app_name(), workload.get_job_type(),
+                      workload.get_thread_count())
+            # as of right now we're only oversubscribing services
+            # TODO expand to batch by checking predicted duration >>> 10m
+            if workload.get_job_type() != SERVICE:
+                continue
+
+            pred = self.__cpu_usage_predictor_manager.get_predictor().predict(workload,
+                                                                              cpu_usage.get(workload.get_id(), None),
+                                                                              pred_env)
+            if pred > self.__config_manager.get_float(TOTAL_THRESHOLD, DEFAULT_TOTAL_THRESHOLD):
+                continue
+            log.debug(' --> low utilization (%f), oversubscribing', pred)
+            free_cpu_count += workload.get_thread_count()
+            workload_count += 1
+
+        if free_cpu_count > 0:
+            self.__add_window(start, end, free_cpu_count)
+
+        self.__success_count += 1
+        self.__reclaimed_cpu_count = free_cpu_count
+        self.handled_event(event, 'oversubscribed {} total cpus from {} workloads'.format(free_cpu_count,
+                                                                                          workload_count))
+
+    def __relevant(self, event):
+        if not event[ACTION] == OVERSUBSCRIBE:
+            self.ignored_event(event, 'not a {} event'.format(OVERSUBSCRIBE))
+            return False
+
+        return True
+
+    # NOTE: this creates the CRD. should not be run and only really exists for testing purposes.
```

Comment:
Might be nice for legibility reasons to put a sample of the CRD here in a comment.

Reply:
added and moved the method to the bottom so it isn't a block of comment in the middle of the file.